Library/Homebrew/utils/github/api.rb

```diff
@@ -62,19 +62,30 @@ def pretty_ratelimit_reset(reset)
 
     # Error when authentication fails.
     class AuthenticationFailedError < Error
-      def initialize(github_message)
+      def initialize(credentials_type, github_message)
         @github_message = github_message
         message = +"GitHub API Error: #{github_message}\n"
-        message << if Homebrew::EnvConfig.github_api_token
+        message << case credentials_type
+        when :github_cli_token
           <<~EOS
-            HOMEBREW_GITHUB_API_TOKEN may be invalid or expired; check:
-              #{Formatter.url("https://github.com/settings/tokens")}
+            Your GitHub CLI login session may be invalid.
+            Refresh it with:
+              gh auth login --hostname github.com
           EOS
-        else
+        when :keychain_username_password
           <<~EOS
             The GitHub credentials in the macOS keychain may be invalid.
             Clear them with:
               printf "protocol=https\\nhost=github.com\\n" | git credential-osxkeychain erase
+          EOS
+        when :env_token
+          <<~EOS
+            HOMEBREW_GITHUB_API_TOKEN may be invalid or expired; check:
+              #{Formatter.url("https://github.com/settings/tokens")}
+          EOS
+        else
+          <<~EOS
+            HOMEBREW_GITHUB_API_TOKEN is unset.
```

Comment:
Yes, I think so. Maybe we just need to throw `MissingAuthenticationError` instead if we fall through to here?

Reply:
The case where this could be hit is if you try to take an action that requires authentication against a publicly visible resource. I'm not sure how often Homebrew might attempt that, but it is worth nothing that `test-bot` was attempting that before https://github.com/Homebrew/brew/pull/15486.

In any case, https://github.com/Homebrew/brew/pull/15485#discussion_r1206425367 was the driver behind adding this; I'm happy to walk it back if consensus changes.

Reply:
Combing over the API docs a bit, it looks like there are only a couple scenarios we should run into:
* [401 when the credentials are invalid](https://docs.github.com/en/rest/overview/resources-in-the-rest-api?apiVersion=2022-11-28#failed-login-limit)
* [403 when no User-Agent is provided](https://docs.github.com/en/rest/overview/resources-in-the-rest-api?apiVersion=2022-11-28#user-agent-required)
* [403 when not enough scope is available](https://docs.github.com/en/rest/overview/resources-in-the-rest-api?apiVersion=2022-11-28#authentication)
* [403 when we go over the ratelimit](https://docs.github.com/en/rest/overview/resources-in-the-rest-api?apiVersion=2022-11-28#exceeding-the-rate-limit)

As the only scenarios outside rate-limiting where we can hit 401/403 happen with credentials, I'll go ahead and re-work this code a bit.

Reply:
In the latest iteration I've just let this case fall through as it doesn't seem like it should happen in the 'real world'; I've also left the `pat_blurb` out for now as it doesn't seem to provide any further useful context when the token is invalid.

If the user clears their token and then hits an auth error again, they will get the PAT blurb which is now more actionable/meaningful.